content/en/docs/use-cases/encrypted-images.md

```diff
@@ -0,0 +1,131 @@
+---
+title: Encrypted images
+date: 2023-01-24
+description: Procedures to encrypt and consume OCI images in a TEE
+categories:
+- use case
+tags:
+- coco-keyprovider
+- images
+- kbs
+---
+
+# Using encrypted images in CoCo
+
+A user might want to bundle sensitive data on an OCI (Docker) image. The image layers should only be accessible within a Trusted Execution Environment (TEE).
+
+The project provides the means to encrypt an image with a symmetric key that is released to the TEE only after successful verification and appraisal in a Remote Attestation process. CoCo infrastructure components within the TEE will transparently decrypt the image layers as they are pulled from a registry without exposing them to the outside.
+
+## Instructions
+
+The following steps require a functional CoCo installation on a Kubernetes cluster. A Key Broker Client (KBC) has to be configured for TEEs to be able to retrieve confidential secrets. We assume `cc_kbc` as a KBC for the CoCo Projects KBS in the following instructions, but image encryption should work with other Key Broker implementations similarly.
+
+Please make sure you have a recent version of (skopeo)[https://github.com/containers/skopeo] (v1.14+) installed locally.
+
+### Encrypt an image
+
+We extend public image with secret data.
+
+```bash
+docker build -t unencrypted - <<EOF
+FROM nginx:stable
+RUN echo "something confidential" > /secret
+EOF
+```
+
+The encryption key needs to be a 32 byte sequence and provided to the encryption step as base64-encoded string.
+
+The key id is a generic resource descriptor used by the Key Broker to look up secrets in its storage. For CoCo's KBS this is composed from three segments: `$repository_name/$resource_type/$resource_tag`
+
+```bash
+head -c 32 /dev/urandom | openssl enc > image_key
+KEY_B64="$(base64 < image_key)"
+KEY_PATH="/default/image_key/nginx"
+KEY_ID="kbs://${KEY_PATH}"
+```
+
+The image encryption logic is invoked in a container. To access the unencrypted image from within the container, skopeo can be used to buffer the image in a directory, which is then made available to the container. Similary, the resulting encrypted image will be put into an output directory.
```

Comment:
You mean by mounting docker secrets into the container and pushing to the target registry directly? 

the container wouldn't be able to create a "local" docker image, since it doesn't have access to the docker daemon (we could also mount some socket I guess, but this presumes docker vs podman ...)

Reply:
Yeah, I think it needed to run a container as daemon using docker, mounting the key file, and involved `ocicrypt` configuration. I am not sure about the docker vs podman difference.

I think this (using a dir) method could be convenient for users who wants to use the container immediately locally. 

But when we create this directory like this do we lose the layers information? (I wanted to test this myself, but currently has docker version problems). Building large docker images could be harder when the disk space is constrained.

Reply:
> But when we create this directory like this do we lose the layers information?

No afaik, this is a valid representation of an OCI image, it should keep all the information 1:1. e.g. a roundtrip from library/nginx => dir => library/nginx should keep everything